notebook/notebook/notebook.py

```diff
@@ -113,201 +143,334 @@ def start_pod(jupyter_token, image):
+            log.info("Issue with watcher")
+            log.error(e)
+            w.stop()
+            break
+
+    ws.close()
+
+
+@app.route('/jupyter/verify/<svc_name>/', methods=['GET'])
+def verify(svc_name: str):
+    access_token = request.cookies.get('access_token')
+
+    if is_falsy_str(access_token):
+        return '', 401
+
+    resp = requests.get(f'{AUTH_GATEWAY}/verify',
+                        headers={'Authorization': f'Bearer {access_token}'})
+
+    if resp.status_code != 200:
+        return '', 401
+
+    user_id = resp.headers.get('User')
+
+    if is_falsy_str(user_id):
+        return '', 401
+
+    k_res = k8s.read_namespaced_service(svc_name, 'default')
+    labels = k_res.metadata.labels
+
+    if labels['user_id'] != UNSAFE_user_id_transform(user_id):
+        return '', 401
+
+    # FIXME: Replace with pod ip
+    # resp = Response('')
+    # resp.headers['IP'] = k_res.spec.cluster_ip
 
-@app.route('/healthcheck')
-def healthcheck():
     return '', 200
 
 
-@app.route('/')
-def root():
-    if 'svc_name' not in session:
-        log.info(f'no svc_name found in session {session.keys()}')
-        return render_template('index.html',
-                               form_action_url=external_url_for('new'),
-                               images=list(WORKER_IMAGES),
-                               default='hail')
-    svc_name = session['svc_name']
-    jupyter_token = session['jupyter_token']
-    log.info('redirecting to ' + external_url_for(f'instance/{svc_name}/?token={jupyter_token}'))
-    return redirect(external_url_for(f'instance/{svc_name}/?token={jupyter_token}'))
-
-
-@app.route('/new', methods=['GET'])
-def new_get():
-    pod_name = session.get('pod_name')
-    svc_name = session.get('svc_name')
-    if pod_name:
-        delete_worker_pod(pod_name, svc_name)
-    session.clear()
-    return redirect(external_url_for('/'))
-
-@app.route('/new', methods=['POST'])
-def new_post():
-    log.info('new received')
-    password = request.form['password']
-    image = request.form['image']
-    if password != PASSWORD or image not in WORKER_IMAGES:
-        return '403 Forbidden', 403
-    jupyter_token = uuid.uuid4().hex  # FIXME: probably should be cryptographically secure
-    svc, pod = start_pod(jupyter_token, WORKER_IMAGES[image])
-    session['svc_name'] = svc.metadata.name
-    session['pod_name'] = pod.metadata.name
-    session['jupyter_token'] = jupyter_token
-    return redirect(external_url_for(f'wait'))
-
-@app.route('/wait', methods=['GET'])
-def wait_webpage():
-    return render_template('wait.html')
-
-@app.route('/auth/<requested_svc_name>')
-def auth(requested_svc_name):
-    approved_svc_name = session.get('svc_name')
-    if approved_svc_name and approved_svc_name == requested_svc_name:
-        return '', 200
-    return '', 403
+@app.route('/jupyter', methods=['GET'])
+def get_notebooks():
+    user_id = request.headers.get('User')
 
+    if is_falsy_str(user_id):
```

Comment:
We trust the `User` header enough to pass it to k8s. This check doesn't seem gainful? Worst case we try to find some pods with user_id `None` or ``. The latter will return no pods. The former will error when we try to transform it.

At the end of the day, we have to get the authentication service right before we can trust any other part of our system (since all of it depends on correct authentication).

Reply:
Just to confirm, you would like me to allow user_id == None here?